[WIP] Refactors the Holodeck to use Map Templates

Kylerace opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->
![Screenshot_158](https://user-images.githubusercontent.com/15794172/95807764-cd71b480-0cbf-11eb-9b08-146b274ede22.png)


## About The Pull Request
Refactors the holodeck to use map templates instead of copy_contents_to, which every maintainer seems to have complaints about. 
Fixes #41485 due to an added GetAllContents() proc for all containers in the spawned list
Fixes #49318 because the holodeck doesnt actually change areas anymore, and holodeck/rec_center has the VALID_TERRITORY flag
Fixes #49673, i dont know if this was fixed in some earlier pr but the behavior it talks about doesnt happen anymore so im counting it

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
The holodeck sims wont take up space in the centcom z level any more (which allows for more possible programs in the future), and map templates are more heavily tested. This is also a chance to future proof the holodeck against bugs if possible. Holodeck also seems more responsive. This should allow for a second custom holodeck in some future ruin as well, e.g. var edited for the offline program to look different and/or a completely different list of programs
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:

refactor: Nanotrasen overhauled much of the holodecks programming so that it didnt have to keep physical copies of the simulations anymore.
code: holodeck uses map templates instead of area_copy
/:cl:
EDIT: a lot of this following paragraph is outdated, the list of atoms created by the holodeck is now returned by InitializeAtoms inside of SSatoms, parsed_map/initTemplateBounds passes an argument to InitializeAtoms that makes it return a list only if the map template itself is set to return a list. initAtom() adds itself to the return list in SSatoms if the subsystem is set to return one (it unsets itself after the template is done, and i wasnt able to find a way to trick SSatoms into adding a non template atom to the list when its not supposed to, but it might be possible).
Right now im trying to cut down the number of for loops that iterate through every turf in the holodeck and/or its contents, as well as making sure that my changes to the vanilla map template and parsed_map code doesnt fuck up something other than holodeck.
 i'll write this more fully tomorrow but basically datum/map_template and datum/parsed map both have an added list 
called created_atoms and a boolean returns_list that defaults to false. when template.load() is called it sets its parsed_map instance to share the same boolean and list, then inside parsed_map/create_atom() it adds the created atom to the list if returns_list is true. the flaws ive found with this approach so far is 
1. this by itself doesnt work for objects that create other objects on initialization (like courtroom lockers), requiring all items in spawned to be checked for any contents. 
2.  is a bit ugly since it requires the new values to be passed down and then up the daisy chain of holodeck -> map_template instance -> parsed_map instance (returns_list is inherited from map_template, list is created here) -> same map_template instance (now has populated list) -> holodeck. However map templates already have a similar daisy chain except in the form of proc arguments instead of variables assigned at creation.


<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Kylerace committed:
Merge remote-tracking branch 'upstream/master'

actions-user committed:
Automatic changelog compile [ci skip]

actions-user committed:
Automatic changelog compile [ci skip]

actions-user committed:
Automatic changelog compile [ci skip]

Kylerace committed:
Merge remote-tracking branch 'upstream/master'

Kylerace committed:
Merge remote-tracking branch 'origin/master'

Kylerace committed:
Adds all holodeck sims as map templates

also adds a survival pod that spawns the lounge sim for testing, will remove

Kylerace committed:
adds holodeck map helpers, also many comments not meant for release

Kylerace committed:
changed generate_program_list to work with map templates

honestly im just committing so i can restart my computer its super fucking slow for some reason

Kylerace committed:
mostly(?) refactors necessary procs for mapload

still doesnt work at this point lmao

Kylerace committed:
Top secret Potato documents

Kylerace committed:
fixes ordering?

Kylerace committed:
the interface works! (mostly)

Kylerace committed:
I dont pop changes when switching branches :sunglasses:

overriding stuff related to parsed_maps and map_templates

Kylerace committed:
spawned works for the most part now

dont tell anyone but really this commit is so i can switch to another branch (shhhhh)

Kylerace committed:
e

Kylerace committed:
for testing

Kylerace committed:
actually updating to master now

Kylerace committed:
Merge remote-tracking branch 'upstream/master'

Kylerace committed:
Merge remote-tracking branch 'upstream/master'

Kylerace committed:
gitkraken crashes when loading large difs'

Kylerace committed:
Merge branch 'master' into refactorholodeck

Kylerace committed:
fixes stuff

Kylerace committed:
Putting some commented out features back in

Kylerace committed:
Merge remote-tracking branch 'upstream/master'

Kylerace committed:
power fully works + baseturfs dont replicate

optimumtact commented:
remove the change log changes you have made and disable changelog CI on your fork of tgstation.

Kylerace committed:
Gets rid of unhelpful holodeck mapping helpers

Kylerace committed:
Gets rid of some more unneeded stuff

LemonInTheDark commented:
You've got some file paths that have been updated, and it looks like you need to rebuild tgui, but this is promising, it's been a long time coming too, godspeed :)

LemonInTheDark commented:
Oh also add more detail to what was changed in the changelog

Fikou commented:
you forgot to delete the holodecks from centcom

Dennok commented on code/modules/holodeck/computer.dm:
```diff
@@ -26,43 +26,50 @@
 		log_mapping("Holodeck computer cannot be in a holodeck, This would cause circular power dependency.")
 		qdel(src)
 		return
+
+	// the following is necessary for power reasons
+	if(!linked || !offline_program)
+		log_world("No matching holodeck area found")
+		qdel(src)
+		return
+
```
I think that `!offline_program` must be in separated check and log something like `Wrong offline_program variable`

Kylerace replied:
fixed

Dennok commented on code/modules/holodeck/computer.dm:
```diff
@@ -26,43 +26,50 @@
-		return
+/obj/machinery/computer/holodeck/LateInitialize()//from here linked is populated and the program list is generated. its also set to load the offline program
+	linked = GLOB.areas_by_type[/area/holodeck/rec_center] //this should make current_area be the actual holodeck offline area object
+	bottom_left = locate(linked.x, linked.y, 2)
```
z = 2 ?
It maybe place here cumputer z?

Kylerace replied:
fixed, now its bottom_left = locate(linked.x, linked.y, src.z)

Dennok commented on code/modules/holodeck/computer.dm:
```diff
@@ -72,7 +79,7 @@
 
 	generate_program_list()
-	load_program(offline_program, FALSE, FALSE)
+	load_program("holodeck_offline")//honestly there isnt a reason to do this as far as im aware
```
until someone want custom offlike holodeck, like space, or plasma burn range or etc.

Kylerace replied:
fixed, now uses the offline_program var again in case someone wants to var edit a custom holodeck

Dennok commented on code/modules/holodeck/computer.dm:
```diff
@@ -290,24 +380,23 @@
 	if(!silent)
 		visible_message("<span class='notice'>[O] fades away!</span>")
-	qdel(O)
+		qdel(O)
+	else
+		qdel(O)
```
What is the reason for this?

Kylerace replied:
I was probably playing it safe, fixed

Dennok commented on code/modules/holodeck/turfs.dm:
```diff
@@ -2,6 +2,7 @@
 	icon_state = "floor"
 	thermal_conductivity = 0
 	flags_1 = NONE
+	var/direction = 2
```
Why not just use dir?

Dennok commented on code/modules/mapping/map_template.dm:
```diff
@@ -107,7 +107,7 @@
 		return
 
-	var/list/border = block(locate(max(T.x-1, 1),			max(T.y-1, 1),			 T.z),
+	var/list/border = block(locate(max(T.x-1, 1),			max(T.y-1, 1),			 T.z),//kyler, this makes a block of turfs between the farthest possible turf corner and the closest?
 							locate(min(T.x+width+1, world.maxx),	min(T.y+height+1, world.maxy), T.z))
```
This one make block of turfs between lower left-1 and upper right+1. outer border

Kylerace replied:
I wasn't really sure what key did but I don't pass program.type to tgui anymore so I changed it to id

Dennok commented on tgui/packages/tgui/interfaces/Holodeck.js:
```diff
@@ -31,12 +31,12 @@ export const Holodeck = (props, context) => {
             <Button
               fluid
-              key={def_program.type}
+              key={def_program.id}
```
This change really matters?

Kylerace committed:
Some fixes for dennoks reviews

Kylerace committed:
chapel courtroom doesnt have null air at parts anymore

Kylerace committed:
removes unnecessary changes and hopefully fix tgui?

Kylerace committed:
Merge remote-tracking branch 'upstream/master'

Kylerace committed:
Merge branch 'master' into refactorholodeck

Kylerace committed:
fixes tgui? v2

Kylerace committed:
Many misc changes, updated templates, got rid of unneeded changes

Kylerace commented:
Fuck the changelog, all my homies hate the changelog

Kylerace committed:
spooky cursed changelog for october ooOOooOOoo

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into refactorholodeck

Kylerace committed:
redoes some changes i made before

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into refactorholodeck

Kylerace committed:
Fixes tgui AGAIN and fixes bug caused by not deleting areas in cc

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into refactorholodeck

Kylerace committed:
Auto stash before merge of "refactorholodeck" and "upstream/master"

Kylerace committed:
Big readability improvements, computers spawned after mapload work now

Kylerace committed:
fixes the thunderdome buttons

Kylerace committed:
Code readability improvements + reordering

Kylerace committed:
adds kobayashi maru to the program list

Kylerace committed:
independent access vars between holodeck computers and programs

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into refactorholodeck

Kylerace committed:
Fixes stupid bullshit conflicts

Dennok commented:
You can place area noop  to templates. To make it not change area.

Kylerace commented:
> You can place area noop to templates. To make it not change area.
So the holodeck should have one area no matter what program is loaded? @Dennok

Kylerace committed:
Changes all templates to have area/template_noop

it never changes area anymore, ill change the code later

Dennok commented:
> > You can place area noop to templates. To make it not change area.
> 
> So the holodeck should have one area no matter what program is loaded? @Dennok

Yes

Kylerace committed:
final readability improvement

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into refactorholodeck

Kylerace committed:
rebuilds tgui, actual final readability improvement

Kylerace committed:
fixes difference between branch maps and master maps

Kylerace committed:
fixes multi holodeck support, nerfs gamer debug program

Kylerace committed:
deletes two unneeded comments

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into refactorholodeck

Kylerace committed:
why the fuck did arcanes economy pr change centcom reeeeeeeeeeeeeeeeeeee

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into refactorholodeck

Kylerace committed:
feex tee gee you aiii

Kylerace committed:
its nerf or nothing

Kylerace committed:
forgot something

Cyberboss commented on code/modules/holodeck/holodeck.dm:
```diff
@@ -0,0 +1,207 @@
+
```
Name this `map_templates.dm` or something else more descriptive

Kylerace replied:
now called holodeck_map_templates and updated comment at the top of computer.dm

Cyberboss commented on code/modules/holodeck/turfs.dm:
```diff
@@ -19,6 +20,43 @@
+	//message_admins("INITIALIZE")
+	//return INITIALIZE_HINT_LATELOAD
+
+/turf/open/floor/holofloor/chapel/LateInitialize()
+	//message_admins("LATE INITIALIZE")
+	//if (direction != 2)
+	//	src.setDir(direction)
+	//. = ..()
```
Delete this proc

Kylerace replied:
done

Kylerace committed:
fixes for cyberboss

Kylerace commented:
need to fix explosion bug, after that its good to go

Kylerace committed:
Murders all single letter vars, fixes explosion bug

AnturK commented on code/modules/holodeck/computer.dm:
```diff
@@ -110,63 +121,277 @@
+	// ruins clogging up memory for the whole round.
+	var/datum/parsed_map/parsed = cached_map || new(file(mappath))
+	cached_map = keep_cached_map ? parsed : null
+	if(!parsed.load(target_turf.x, target_turf.y, target_turf.z, cropMap=TRUE, no_changeturf=(SSatoms.initialized == INITIALIZATION_INSSATOMS), placeOnTop=FALSE))
```
If this is the only change it would be better to make this variable on map_template type.

Kylerace replied:
removed snowflake holodeck map loader procs, now all map_templates (and parsed_map) instances have a list and boolean var that defaults to false, if its set to true then parsed_map/proc/create_atom() adds the created element to the list.

AnturK commented on code/modules/holodeck/computer.dm:
```diff
@@ -110,63 +121,277 @@
+/*
+	similar to initTemplateBounds except it keeps track of all new objects it creates and passes it to map_template/holodeck/load
+*/
+/datum/parsed_map/proc/holodeckTemplateBounds()
```
Same thing here. Duplication here is bad.

Kylerace replied:
no more duplication

AnturK commented on code/modules/holodeck/computer.dm:
```diff
@@ -110,63 +121,277 @@
+
+	spawned_atoms = parsed.holodeckTemplateBounds()
+
+	for (var/obj/base_container in spawned_atoms)//this didnt seem to work when i put this in holodeckTemplateBounds
```
Is this necessary if you already iterate them one by one during template initialization.

Kylerace replied:
this is for lockers that create other items on initialization, thus the insides arent actually spawned from the template which makes them untracked otherwise

Kylerace replied:
ill probably end up putting this part into load_program

Kylerace replied:
leaving this open because its possible theres a better way to accomplish this without manually adding all contents of every spawned container just in case they create items on Initialize()

Kylerace replied:
as of right now this is fixed unless i get smacked for altering ssatoms to return a list of atoms created from a template (if the template is set to return a list, which only holodeck templates are currently)

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into refactorholodeck

Kylerace committed:
wip: removes snowflake load procs and adds functionality to map template

Kylerace committed:
wip still, clean up and need to fix bugs with power toggling

Kylerace committed:
mass fixes, added turf blacklists to map templates

Kylerace committed:
power and closed wall in holodeck issues are now FIXED

Kylerace committed:
sorry commit number 69, i gotta switch branches D:

LemonInTheDark commented:
need more review?

Kylerace committed:
got rid of an unnecessary argument in build_coordinate

Kylerace commented:
> need more review?

@LemonInTheDark yes reviews would help a lot. specifically I need to know if what im doing in map_template.dm and reader.dm (plus the map_template/holodeck/update_blacklist() proc in line 135 of computer.dm) is ok. I tested it and on average my branch and master are done with creating the station + lavaland at nearly the same time. if im right non holodeck map templates shouldnt be doing any more work than evaluating some if statements to be false and calling a proc that just returns

Kylerace committed:
Merge branch 'master' into refactorholodeck

Kylerace committed:
fixes tgui

Kylerace committed:
updates comment about turf dir bug

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into refactorholodeck

Kylerace committed:
fixes tgui and power toggling memes

Kylerace committed:
first step changes to adding list return functionality to SSatoms

helps get rid of snowflake shit from holodeck and lets other things use it in the future

Kylerace committed:
shitcode ELIMINATED

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into refactorholodeck

Kylerace committed:
tgui thingy

Kylerace committed:
gets rid of var/ in proc arguments because of the big mean linter

AnturK commented:
How's this going.

Kylerace commented:
@AnturK ive decided to shelve the holo element idea since it doesnt help enough to be worth it. Might return to it in a future pr though. Currently I don't see anything else to do with it unless a maintainer wants me to change something.

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into refactorholodeck

Kylerace committed:
fixes tgui

Kylerace committed:
gets rid of a redundant for loop in emergency_shutdown()

AnturK commented on code/modules/holodeck/computer.dm:
```diff
@@ -110,63 +120,193 @@
+			var/program_to_load = params["id"]
+			//load the map_template that program_to_load represents
+			if(program_to_load)
+				load_program(program_to_load)
```
You seem to have dropped the check for if the template actually exists and if emagged template is loaded into actually emagged holodeck.

AnturK commented on code/modules/holodeck/computer.dm:
```diff
@@ -110,63 +120,193 @@
+		if (istype(holo_turf, /turf/closed))
+			for (var/_baseturf in holo_turf.baseturfs)
+				if (ispath(_baseturf, /turf/open/floor/holofloor))
+					holo_turf.baseturfs -= _baseturf
+					holo_turf.baseturfs += /turf/open/floor/holofloor/plating
```
Description in comments what's happening here please.

AnturK commented on code/modules/holodeck/computer.dm:
```diff
@@ -110,63 +120,193 @@
+	template = SSmapping.holodeck_templates[map_id]
+	for (var/_turf in linked)
+		var/turf/holo_turf = _turf
+		if (istype(holo_turf, /turf/closed))
```
If you're doing istype here you can just do typed loop, it's more readable.

AnturK commented on code/modules/holodeck/computer.dm:
```diff
@@ -110,63 +120,193 @@
+
+		atoms.flags_1 |= HOLOGRAM_1
+
+		if (istype(atoms, /obj/effect/holodeck_effect/))//this is what makes holoeffects work
```
Would be nice to wrap these holoeffects into some typed proc while you're doing this.

AnturK commented on code/modules/holodeck/computer.dm:
```diff
@@ -110,63 +120,193 @@
+	spawned -= object
+	var/turf/target_turf = get_turf(object)
+	for(var/atom/movable/object_contents in object) // these should be derezed if they were generated
+		object_contents.forceMove(target_turf)
```
Why is this move necessary here ?

Kylerace replied:
leftover from original holodeck code that i didnt get rid of, ill take it out in the next update (assuming it doesnt ruin anything but i dont see why it would)

Kylerace replied:
found out why its necessary, if a non holodeck item is inside of a holodeck container (like the morgue trays), they can get qdel'd with the container without this proc. so far it only seems to work like that with the morgue tray, I assume that closets do it themselves on deletion.

AnturK commented on code/modules/mapping/reader.dm:
```diff
@@ -381,6 +391,8 @@
 /datum/parsed_map/proc/create_atom(path, crds)
 	set waitfor = FALSE
 	. = new path (crds)
+	//if (returns_created)
```
Remains here and in vars

AnturK reviewed (changes requested):
Looks mostly okay, a few minor issues. Also needs docs.

Kylerace committed:
fixes tons of shit anturk wants

Kylerace committed:
initTemplateBounds now a map_template proc, cuts down unneeded code

Kylerace commented:
definitely dont merge yet, have consistent oom crashing issues since one or two commits ago

Kylerace commented:
havent been able to make real progress on this in days, verifying that a given version doesnt cause occasional crashes takes forever. I have to shelve this until next quarter in two weeks or so, i have too many things going on right now unfortunately.
